### PR TITLE
[Feature]Support reorder ids to split prefill and decodes 

### PR DESCRIPTION
**Pull request by bukejiyu**
## Motivation

PR from https://github.com/PaddlePaddle/FastDeploy/pull/5194
为了能够更好的接入三方Attention，需要对输入进行重排，将prefill token和decode token区分开来，本PR支持了重排功能，目前支持了基础场景及投机解码场景下的重排

## Modifications
**多模模型 仅在V1调度下才支持 PD重排 export ENABLE_V1_KVCACHE_SCHEDULER=1**
**当前PD重排仅支持 CUDA backend**
1.新增InputBatch结构用于gpumodelrunner share_input 管理,新增ProposerInputBatch 用于mtp share_input 管理，用于管理gpu_model_runner的输入，并且添加reorder_split_prefill_and_decode和condense函数支持重排
2.merge develop
3.为每个VL请求增加req_id -> img_features的映射方便重排

## Usage or Command

在 AttentionBackend 中将类变量 enable_ids_reorder （默认为False）设为 True，即可启用 P/D 重排

## Checklist

- [x] Add at least a tag in the PR title.
  - Tag list: [`[FDConfig]`,`[APIServer]`,`[Engine]`, `[Scheduler]`, `[PD Disaggregation]`, `[Executor]`, `[Graph Optimization]`, `[Speculative Decoding]`, `[RL]`, `[Models]`, `[Quantization]`, `[Loader]`, `[OP]`, `[KVCache]`, `[DataProcessor]`, `[BugFix]`, `[Docs]`, `[CI]`, `[Optimization]`, `[Feature]`, `[Benchmark]`, `[Others]`, `[XPU]`, `[HPU]`, `[GCU]`, `[DCU]`, `[Iluvatar]`, `[Metax]`]
  - You can add new tags based on the PR content, but the semantics must be clear.
- [x] Format your code, run `pre-commit` before commit.
- [x] Add unit tests. Please write the reason in this PR if no unit tests.
- [ ] Provide accuracy results.
- [ ] If the current PR is submitting to the `release` branch, make sure the PR has been submitted to the `develop` branch, then cherry-pick it to the `release` branch with the `[Cherry-Pick]` PR tag.

### CODE REVIEW

File: fastdeploy/spec_decode/mtp.py [hunk 1]
```diff
@@ -169,7 +176,7 @@ def dummy_prefill_inputs(self, num_tokens: int, batch_size: int, expected_decode
             idx = i
             self.model_inputs["input_ids"][idx : idx + 1, :input_length] = np.array([5] * input_length)
             self.model_inputs["eos_token_id"][:] = np.array([2], dtype="int64").reshape(-1, 1)
-            self.seq_lens_this_time_buffer[idx : idx + 1] = input_length
+            self.model_inputs["seq_lens_this_time_buffer"][idx : idx + 1] = input_length
```
Comment (Deleter-D) [inline fastdeploy/spec_decode/mtp.py:183]: 这里model_inputs已经是一个object了，还并存了dict的key-value访问方式是否合理？
原本此处逻辑seq_lens_this_time_buffer是MTPProposer的成员变量，现在又合并回了model_inputs里，是否有其他影响？

File: fastdeploy/spec_decode/mtp.py [hunk 2]
```diff
@@ -415,175 +422,25 @@ def update_mtp_block_num(self, num_gpu_blocks) -> None:
...
-            )
-
     def insert_tasks_v1(self, req_dicts: List[Request], num_running_requests: int):
 
         if "caches" not in self.model_inputs:
             self.initialize_kv_cache()
         req_len = len(req_dicts)
-
+        self.model_inputs["num_running_requests"] = num_running_requests
+        self.model_inputs["running_requests_ids"] = range(num_running_requests)
```
Comment (Deleter-D) [inline fastdeploy/spec_decode/mtp.py:456]: 同上

File: fastdeploy/spec_decode/mtp.py [hunk 3]
```diff
@@ -639,15 +496,15 @@ def insert_tasks_v1(self, req_dicts: List[Request], num_running_requests: int):
...
                 self.model_inputs["seq_lens_decoder"][idx : idx + 1] = 0
                 self.model_inputs["seq_lens_encoder"][idx : idx + 1] = 0
                 self.model_inputs["is_block_step"][idx : idx + 1] = False
                 continue
 
         # TODO(liuzichang): Solve splitewise-p bug to restore
-        # self.model_inputs["seq_lens_this_time"] = self.seq_lens_this_time_buffer[:num_running_requests]
-        self.model_inputs["seq_lens_this_time"] = self.seq_lens_this_time_buffer
+        # self.model_inputs["seq_lens_this_time"] = self.model_inputs["seq_lens_this_time_buffer"][:num_running_requests]
+        self.model_inputs.seq_lens_this_time = self.model_inputs["seq_lens_this_time_buffer"]
```
Comment (Deleter-D) [inline fastdeploy/spec_decode/mtp.py:532]: 同上

File: fastdeploy/worker/gpu_model_runner.py [hunk 4]
```diff
@@ -351,7 +352,7 @@ def _init_speculative_proposer(self):
         if self.speculative_method == "ngram":
             self.proposer = NgramProposer(self.fd_config)
         elif self.speculative_method == "mtp":
-            self.share_inputs["seq_lens_this_time"] = self.seq_lens_this_time_buffer
+            self.share_inputs["seq_lens_this_time"] = self.share_inputs["seq_lens_this_time_buffer"]
```
Comment (Deleter-D) [inline fastdeploy/worker/gpu_model_runner.py:378]: 与MTPProposer中的同问

File: fastdeploy/spec_decode/mtp.py (hunk 1 above)
Comment (bukejiyu) [reply fastdeploy/spec_decode/mtp.py:183]: 改动的地方太多了，保留了用key访问的接口，gpumodelrunner用的 InputBatch ，mtp的对象是 ProposerInputBatch 都是自己独立的 MTPProposer的成员变量 和 ProposerInputBatch 的成员变量应该就只是 多套了一层的差别吧，seq_lens_this_time_buffer也是和 req id强相关的只能放到 InputBatch内部才能参与排序

**Comment by CSWYF3634076** [comment]
代码行数比较多，单测覆盖率需要补齐，尤其是input_batch.py

File: fastdeploy/worker/gpu_model_runner.py [hunk 5]
```diff
@@ -550,28 +573,31 @@ def _process_mm_features(self, request_list: List[Request]):
...
+
+                    feature_start = feature_position.offset
+                    feature_end = feature_position.offset + feature_position.length
+                    merge_image_features.append(mm_feature[feature_start:feature_end])
+                    feature_idx += mm_token_lenght
+                    thw_idx += 1
+                image_features_list.append(paddle.concat(merge_image_features, axis=0))
+            for _, index in req_idx_img_index_map.items():
+                if index != -1:
+                    self.share_inputs["image_features_list"][idx] = image_features_list[index]
```
Comment (CSWYF3634076) [inline fastdeploy/worker/gpu_model_runner.py:631]: 请教下上次shape对不上是因为这里只for了一次吗，还是其他问题

File: fastdeploy/worker/gpu_model_runner.py [hunk 6]
```diff
@@ -512,14 +523,26 @@ def _process_mm_features(self, request_list: List[Request]):
...
                     )
+                    multi_vision_inputs["feature_position_list_batches"].append(
+                        self._get_feature_positions(
+                            mm_positions=inputs["mm_positions"][request.num_image_start : request.num_image_end],
+                            prefill_start_index=request.prefill_start_index,
+                            prefill_end_index=request.prefill_end_index,
+                        )
+                    )
 
         if self.encoder_cache is not None:
```
Comment (kevincheng2) [inline fastdeploy/worker/gpu_model_runner.py:575]: 同上

File: fastdeploy/worker/gpu_model_runner.py [hunk 7]
```diff
@@ -435,6 +441,8 @@ def _process_mm_features(self, request_list: List[Request]):
                 rope_3d_position_ids["max_tokens_lst"].append(request.get("max_tokens", 2048))
 
             if request.with_image:
+                req_idx_img_index_map[idx] = img_index
+                img_index = img_index + 1
                 inputs = request.multimodal_inputs
                 if self.encoder_cache is not None:
                     if envs.FD_ENABLE_MAX_PREFILL:
```
Comment (kevincheng2) [inline fastdeploy/worker/gpu_model_runner.py:471]: 这里使用 encoder_cache 的场景感觉也需要 feature_position_list_batches 记录每条请求的位置信息？

**Review by kevincheng2 — CHANGES_REQUESTED** [review]

**Comment by bukejiyu** [comment]
> 代码行数比较多，单测覆盖率需要补齐，尤其是input_batch.py

有的 新增了单测，目前覆盖率没有过的地方是 get_attention_meta这个函数没有命中

File: fastdeploy/worker/gpu_model_runner.py (hunk 5 above)
Comment (bukejiyu) [reply fastdeploy/worker/gpu_model_runner.py:631]: 对可以理解成只for了一次，并且以前视频输入 也没有和req_id 绑定 导致重排很困难，目前是用新的list和req_id绑定上，每次append 进 image_features_list 都是某一个req_id 的图像特征

File: fastdeploy/worker/gpu_model_runner.py (hunk 7 above)
Comment (bukejiyu) [reply fastdeploy/worker/gpu_model_runner.py:471]: 这个我不太确定也，我看他都是用append的应该是自带位置信息的吧？
edf11c35add0c40eaf6733132feeb064

File: fastdeploy/worker/gpu_model_runner.py
Comment (ST-XX) [inline fastdeploy/worker/gpu_model_runner.py:1109]: 这里的 idx 会不会变化呢，这里是 P 阶段的时候会发起受限解码的异步 compile

Comment (bukejiyu) [reply fastdeploy/worker/gpu_model_runner.py:1109]: 会变化，可以把受限解码内部和槽位强相关的tensor 放入input_batch内管理？

File: tests/e2e/test_EB_Lite_serving.py [hunk 8]
```diff
@@ -1417,10 +1417,9 @@ def test_profile_reset_block_num():
         pytest.fail(f"Invalid number format: {match.group(1)}")
 
     lower_bound = baseline * (1 - 0.05)
-    upper_bound = baseline * (1 + 0.05)
     print(f"Reset total_block_num: {actual_value}. baseline: {baseline}")
 
-    assert lower_bound <= actual_value <= upper_bound, (
-        f"Reset total_block_num {actual_value} 与 baseline {baseline} diff需要在5%以内"
-        f"Allowed range: [{lower_bound:.1f}, {upper_bound:.1f}]"
+    assert actual_value >= lower_bound, (
```
Comment (zoooo0820) [inline tests/e2e/test_EB_Lite_serving.py:1422]: 这里为什么要去掉upper_bound的检查呢

File: fastdeploy/worker/input_batch.py [hunk 9]
```diff
@@ -0,0 +1,828 @@
...
+
+        def swap_data(tensor, idx1, idx2):
+            """Safely swap tensor slices using clone"""
+            temp = tensor[idx1].clone()
+            tensor[idx1] = tensor[idx2].clone()
+            tensor[idx2] = temp
+
+        self.index_to_batch_id[i1], self.index_to_batch_id[i2] = self.index_to_batch_id[i2], self.index_to_batch_id[i1]
+        swap_data(self.pre_ids, i1, i2)
+        swap_data(self.input_ids, i1, i2)
```
Comment (zoooo0820) [inline fastdeploy/worker/input_batch.py:329]: 这里以后是否可以考虑用自动遍历成员的方式写，需要swap的太多了，感觉很容易遗漏，另一方面后面如果扩展了其他输入，也容易忘记加

File: fastdeploy/worker/input_batch.py [hunk 10]
```diff
@@ -0,0 +1,828 @@
...
+from paddleformers.utils.log import logger
+
+from fastdeploy.config import CacheConfig, FDConfig, ModelConfig, SpeculativeConfig
+from fastdeploy.model_executor.layers.rotary_embedding import get_rope
+from fastdeploy.model_executor.logits_processor import build_logits_processors
+
+
+class InputBatch:
+    def __getitem__(self, key):
+        """Support dictionary-style attribute access"""
```
Comment (zoooo0820) [inline fastdeploy/worker/input_batch.py:27]: 目前这个类是直接给GPU使用吗，后续其他硬件接入扩展的时候需要怎么引用以及修改

File: fastdeploy/worker/input_batch.py (hunk 9 above)
Comment (bukejiyu) [reply fastdeploy/worker/input_batch.py:329]: 不太好整 有一些属性是 mtp模型 和 target模型共享的，而且也有一些属性和 req id 无关的 没法排序

File: fastdeploy/worker/input_batch.py (hunk 10 above)
Comment (bukejiyu) [reply fastdeploy/worker/input_batch.py:27]: 其他硬件得维护一个新类了

**Review by zhoutianzi666 — APPROVED** [review]

**Review by zoooo0820 — APPROVED** [review]

**Review by Deleter-D — APPROVED** [review]
